sabnzbd/dirscanner.py

```diff
@@ -124,90 +119,116 @@ def run(self):
             if self.dirscan_speed and not self.shutdown:
                 self.scan()
 
-    def scan(self):
-        """Do one scan of the watched folder"""
+    def __listfiles(self, folder, catdir=None):
+        """Generator listing possible paths to NZB files"""
 
-        def run_dir(folder, catdir):
-            try:
-                files = os.listdir(folder)
-            except OSError:
-                if not self.error_reported and not catdir:
-                    logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(folder))
-                    self.error_reported = True
-                files = []
-
-            for filename in files:
-                if self.shutdown:
-                    break
-                path = os.path.join(folder, filename)
-                if os.path.isdir(path) or path in self.ignored or filename[0] == ".":
-                    continue
-
-                if filesystem.get_ext(path) in VALID_NZB_FILES + VALID_ARCHIVES:
-                    try:
-                        stat_tuple = os.stat(path)
-                    except OSError:
-                        continue
-                else:
-                    self.ignored[path] = 1
-                    continue
+        cats = config.get_categories() if catdir is None else {}
+
+        try:
+            with os.scandir(os.path.join(folder, catdir or "")) as it:
+                for entry in it:
+                    if self.shutdown:
+                        break
 
-                if path in self.suspected:
-                    if compare_stat_tuple(self.suspected[path], stat_tuple):
-                        # Suspected file still has the same attributes
+                    path = entry.path
+
+                    if path in self.ignored:
+                        # We still need to know that an ignored file is still present when we clean up
+                        yield path, catdir, None
                         continue
-                    else:
-                        del self.suspected[path]
 
-                if stat_tuple.st_size > 0:
-                    logging.info("Trying to import %s", path)
+                    # If the entry is a catdir then recursion
+                    if entry.is_dir():
+                        if not catdir and entry.name.lower() in cats:
```

Comment:
Surprised this hasn't caused problems for users, as far as I know the values of `cats` are not guaranteed to be lowercase.. But the old code also doesn't seem to care so I guess it's fine..